Make __getitem__ TINYer

geohotstan opened the pull request:
cleean refactooooooooooooors only (maybe)
still potential lines to be reduced in fancy indexing.
wip
wip
wip

geohot commented:
Are all the exception cases tested?

tinyb0t commented:
Changes made in `tinygrad/`:
```
------------------------------------------------------------
files                             insertions       deletions
------------------------------------------------------------
tinygrad/tensor.py                        13              21
------------------------------------------------------------
lines added in the tinygrad folder: -8
```

geohotstan commented:
> Are all the exception cases tested?

I didn't delete any of the exception cases but I did move them around a little. :s
I'll check and if they're not tested, I'll add tests.

geohot commented:
Does GPT-4 have any good ideas?

https://chat.openai.com/share/30ab0b7e-d72c-444f-a228-620b132e9500

geohotstan commented:
idk if collapsing the reshapes into 1 liners was the right choice for fancy indexing since we added 1 extra loop as well as we now have conditional checks within the loops (which could be mitigated with a `list + list` but then that's quite a long 1 liner...

and line 321 is totally unreadable

I added quite a lot of comments here... I'm not sure if it makes it more clear

wozeparrot commented on tinygrad/tensor.py:
```diff
@@ -265,78 +265,66 @@ def pad(self, arg: Tuple[Tuple[int, int], ...], value:float=0) -> Tensor:
-        ret = ret.permute(order=order)
+      sum_dim = [d if n==0 else d+max_dim-n for n,d in enumerate(dim)]
+      arange = [Tensor.arange(ret.shape[d], dtype=dtypes.int32, requires_grad=False, device=self.device).reshape(*[1]*sd, ret.shape[d], *[1]*(((ret.ndim + max_dim) if n==0 else (ret.ndim + max_dim - n)) - sd - 1)) for n,(sd,d) in enumerate(zip(sum_dim, dim))]
+      idx = [i.reshape(*[1]*sum_dim[0], *((1, *[1]*(max_dim-i.ndim), *i.shape, *[1]*(ret.ndim - sd - 1)) if n == 0 else (*[1]*(max_dim-i.ndim), *i.shape, *[1]*(ret.ndim - sum_dim[0] - n)))) for n,(sd,i) in enumerate(zip(sum_dim, idx))]
```
this is pretty unreadable, can you split this line up a bit?

geohotstan replied:
Really happy with the new arange change. 
Current arange is readable and I separated idx 1 liner to 3 lines. Probably more readable

geohot commented on test/test_ops.py:
```diff
@@ -574,11 +574,16 @@ def test_slice_in_bounds_multidim(self):
     helper_test_op([(3,3,3)], lambda x: x[1:2, 1:2, 0:-1], lambda x: x[1:2, 1:2, 0:-1])
 
-  def test_slice_with_none(self):
+  def test_slice_with_none_and_npnewaxis(self):
```
Why is np.newaxis supported? This should be the same as None

geohotstan replied:
`  # - Indexing with np.newaxis or None on a given axis will add a new dimension of size one before that axis`
this line in the description of `_getitem_` mentioned npnewaxis and I didn't see it tested xD

But `np.newaxis` just evaluates to `None` when passed in as arg to `_getitem_`, so really they literally are the same thing.
Yeah, you're right, I'll also del the np.newaxis in the description. 
People should just use None instead.

geohot commented:
This good to merge?

geohotstan commented:
> This good to merge?

good on my end :D
